accepted/0000-link-packages-to-source-and-build.md

```diff
@@ -0,0 +1,514 @@
+
+In November of last year alone, we saw three supply chain attacks against popular npm packages ([UA-Parser-JS](https://www.esentire.com/security-advisories/npm-library-supply-chain-attack), [COA](https://www.esentire.com/security-advisories/coa-npm-supply-chain-attack), and [RC](https://www.esentire.com/security-advisories/rc-npm-supply-chain-attack)), each with millions of downloads per month.
+
+The following chart illustrates several attack vectors facing the supply chain:
+
+![supply-chain-threats](https://user-images.githubusercontent.com/20165/174750708-2be483ac-7e41-4bc3-8ee9-440ef33d9423.svg)
+
+Our focus is on mitigating package hijacking attacks where packages are built from a modified source (C) in the above diagram. This is where a malicious version of an existing open source package gets uploaded to the registry. These maliciously crafted versions are rarely reflected in the canonical source code repository and uploaded directly from an attacker's machine.
+
+The [above](https://www.esentire.com/security-advisories/npm-library-supply-chain-attack) [attacks](https://www.esentire.com/security-advisories/coa-npm-supply-chain-attack) [are](https://www.esentire.com/security-advisories/rc-npm-supply-chain-attack) examples of this, and they frequently occur due to compromised npm credentials but can also happen due to compromised CI/CD or builds.
+
+Code on GitHub, GitLab etc can be browsed and audited, but packages in the registry are opaque and much harder to vet. If packages are built and published in the open these attacks become a whole lot harder. However, right now there’s no way of knowing where a package came from when you retrieve it from the registry.
+
+We want to solve the problem of npm packages being disconnected from their source by linking the published package to the source code repository and build that it originated from.
+
+Once a package includes provenance information (where and how it was built) we can start showing this when browsing a package on the npm registry. You would be able to click through to the particular commit and build that published a given version:
+
+<img width="376" alt="screenshot of npm package with provenance information" src="https://user-images.githubusercontent.com/20165/176637201-fdb02c11-c810-48e1-a203-ea4fb2008ad3.png">
+
+Developers consuming open source packages should get the benefit of this without any changes to their workflows. To begin with, the package integrity should be verified when running the  `npm audit signatures` command and eventually transparently integrated into `npm install` and enabled by default.
```

Comment:
Indeed; the entire point of `npm audit foo` when we discussed `npm audit signatures` was that every `foo` would be independently configurable and invokable.

Reply:
> existing `npm audit signatures` functionality and the new functionality as independent features

Thoughts on verifying all available signatures (registry and build) when running `npm audit signatures` but allowing you to filter the type with an argument like: `npm audit signatures --type=registry|build` (naming tbd)?

Reply:
There is going to be broader shared behavior between the various signature verification commands such as "what do you do if packages are not yet in cache" that lead me to feeling a single command makes far more sense

Reply:
FWIW I'm not at all against a single command but I have seen so many situations where people want *just one task* and don't care about other tasks in a bundled command.

It feels... very un-npm-y to me not to be able to opt-in to choosing more limited but precise sets of work that do the exact thing I want completed, especially when the functionality is fundamentally not related (checking what you received is what's on npm vs. what this RFC proposes).

Reply:
Also to answer @feelepxyz: that's a totally fine solution to me!

Reply:
`npm audit` is intended to be the single command - every subpart of it is intended to be granularly configurable and disableable by users, so that nobody is forced to do a check they don't want.

Reply:
If the consensus is to split it up into two separate functions, I would recommend splitting something like this:

```sh
npm audit signatures # runs everything related to signatures
npm audit signatures compare # only computes the hashes and compares
npm audit signatures integrity # validates the signature and ensures it is present and trusted in sigstore
```

This way, `npm audit` remains a single command, and `npm audit signatures` does not result in a user misunderstanding what it does.